pyproject.toml

```diff
@@ -29,3 +62,111 @@ ignore_missing_imports = true
 disallow_untyped_calls = true
 disallow_untyped_defs = true
 disallow_incomplete_defs = true
+
+[tool.poetry.dependencies]
+python = ">=3.7,<3.11"
+pydicom = "^2.3.0"
+typing-extensions = "*"
+setuptools = "*"
+
+# ----- #
+# Apps  #
+# ----- #
+sqlalchemy = { version =">=2.0", optional = true } # groups ["all", "tests", "dev" ]
+
+# ------------- #
+# Documentation #
+# ------------- #
+sphinx = { version = "5.3.0", optional = true } # groups = ["all", "docs"]
+sphinx-argparse = { version = "*", optional = true } # groups = ["all", "docs"]
+sphinx_copybutton = { version = "*", optional = true } # groups = [ "all", "docs"]
+sphinx_rtd_theme = { version = "*", optional = true } # groups = [ "all", "docs"]
+# sphinx-ext-napoleon = { version = "*", optional = true }  # groups = ["all", "docs"]
+# sphinx-book-theme = { version = "*", optional = true }   # groups = ["all", "docs", "doctests"]
+# jupyter-book = { version = ">=0.12.0", optional = true } # groups = ["all", "docs"]
+# networkx = { version = "*", optional = true }            # groups = ["all", "docs"]
+
+# ----- #
+# Tests #
+# ----- #
+pytest = { version = "*", optional = true }                 # groups = ["all", "tests"]
+pytest-sugar = { version = "*", optional = true }           # groups = ["all", "tests"]
+# hypothesis = { version = "<6", optional = true }          # groups = ["all", "tests"]
+# astroid = { version = "*", optional = true }              # groups = ["all", "tests"]
+# psutil = { version = "*", optional = true }               # groups = ["all", "tests"]
+pylint = { version = "*", optional = true }                 # groups = ["all", "tests"]
+pytest-rerunfailures = { version = "*", optional = true }   # groups = ["all", "tests"]
```

Comment:
Not the ones I commented out, so I should remove those, even though they have no impact because they are commented out.
The ones not commented out are used by CI/CD so the developer needs them to perform those tasks before committing

Reply:
Ok, thanks. I was only referring to `pytest-rerunfailures` (and `pytest-sugar` before), as these also should not be needed for local tests. 
As for the commented out things - yes, you can just remove them (though `astroid` is need by `pylint`, as far as I know...).

Reply:
Aa a developer running pytest, having the ability to only rerun failures is important for a tight loop on fixing something.  I suppose I could also just specify the particular tests involved but rerun failures pretty much automates it.

Reply:
Sure. Generally, I like to have a configuration that allows to run the test locally the same way as in the CI, and I understand that in this case there will be a few dependencies that are not needed for CI, but very helpful locally. I usually have my local venv that I use for testing that has what I need, but this way it is more consistent.

On a related note (not directly related to this PR), I think it would help to use the same approach for all `pydicom` core repositories (by that I mean the ones closely related to `pydicom`, e.g. `pydicom` itself, `pynetdicom`, `pylibjpeg` and the various `pylibjpeg` plugins, and `pyjpegls`). I would like to hear what @darcymason thinks on this account - he is currently working on restructuring `pydicom` for the eventual 3.0 release, and this is something that may also be considered. The tests/checks are currently a bit of a patchwork, created by PRs by different people, and I would like to see a more unified approach for the mentioned repos. Maybe we can create a separate issue in `pydicom` for this...

Reply:
Ah...  One could have a separate group that is for the automated CI/CD, if that level of efficiency is important. More or less same as tests, but cut out the stuff that is for local / development work.
@SimonBiggs has done fantastic work on the automation of pymedphys, and that automation makes maintenance (and being a maintainer) much easier.

But one step at a time.
Let's fix up the things that were effectively broken, and get enough of a first step so that a new pynetdicom contributer such as myself can start on close to level ground.

I want to try to keep my PRs as small as practical.  Even if I don't always succeed.

Reply:
> But one step at a time.

Yes - I didn't mean this for this PR, it was just a convenient place to put it. This should ultimately handled separately, should we decide to do it.

And yes, it's all about making maintenance and being a maintainer easier, given the limited amount of these 😄

Reply:
> I would like to hear what @darcymason thinks on this account - he is currently working on restructuring `pydicom` for the eventual 3.0 release, and this is something that may also be considered. The tests/checks are currently a bit of a patchwork, created by PRs by different people, and I would like to see a more unified approach for the mentioned repos. Maybe we can create a separate issue in `pydicom` for this...

Agree about the patchwork, and that is why I have been spending the time migrating to pyproject.toml-based config (clearly the future now) and trying to do so in a build-backend-independent way as much as possible.  As mentioned in another issue somewhere, I've tried poetry several times over the years and never really got it.  And I get the impression that a lot of its functionality has now been moved into pyproject.toml ways that are compatible with other tools.  All [the build tools are competing now](https://chriswarrick.com/blog/2023/01/15/how-to-improve-python-packaging/) and this space will be a bit messy for a while.

As to running locally, there is [act](https://github.com/nektos/act) which actually picks up the github actions workflows.  I looked at it (or something similar) in the past but I think it didn't have Windows support previously.  In any case, it appeals to me because it would be nice to just directly run the github actions.  I intend to give it a try soon.

And to continue to be a bit contrarian :wink: I prefer _not_ pinning versions of tools, I'd rather see the error come up and try to deal with it.  I recently got over the pydicom `mypy` pinning - while refactoring anyway, I couldn't remember exactly why it had been pinned, and thought it better to fix up a bunch of problems with the current version, rather than some previous version and then later try to upgrade versions.

In any case, I think `act` (or similar tool) is likely a permanent solution for CI vs local, if it works.

Reply:
> and that is why I have been spending the time migrating to pyproject.toml-based config (clearly the future now) and trying to do so in a build-backend-independent way as much as possible

Yes, using `pyproject.toml` is now the way to go, though that leaves the question of which tools to use open. I had a try at `poetry` several years ago and IIRC it didn't impress me, but I guess this has improved together with the whole packaging metaverse. I'm fairly agnostic concerning the used tools, I just want to have some consistence for ease of maintenance. This includes the "meta tools" like petry, and the actual linters like flake8 etc - would be good to have a single approach. Also, we currently have both command line tools, and GH integrations via annotations, and I tend to get confused with all the different linters.


>  I recently got over the pydicom mypy pinning - while refactoring anyway, I couldn't remember exactly why it had been pinned

The reason was that the build usually failed in some random PR after a `mypy` update, and every time this happened it created some back-and-forth with the PR creator as to how to handle these errors, which are completely unrelated to the PR. Also, this used to happen in several PRs at once, so we had to decide, in which one it had to be fixed.
As I wrote, I would like to have both: a pinned version for the PRs, so that the creator of the PR can concentrate on their own changes, and an unpinned version which shows the errors, but does not fail the build. As I understood, `pymedphys` does something like this, though I may have got that wrong.

> In any case, I think act (or similar tool) is likely a permanent solution for CI vs local, if it works.

Independently of what I wrote before, this sounds like a good thing. It also seems to work under Windows now (using Docker Desktop). `Pre-commit` is also an option - I use it in another repo, though that runs only for the current OS (in an isolated environment).

Reply:
> This includes the "meta tools" like petry, and the actual linters like flake8 etc - would be good to have a single approach.

Of note on the latter, I was really impressed with `ruff` now that switching to it was contributed in pydicom.  Super-fast, and covers everything, including (as in my last PR) upgrading for Python version changes.  Mind you, it still uses all the flags from the previous linters, so some of the issues might still follow.

> The reason was that the build usually failed in some random PR after a `mypy` update, and every time this happened it created some back-and-forth with the PR creator as to how to handle these errors, which are completely unrelated to the PR. Also, this used to happen in several PRs at once, so we had to decide, in which one it had to be fixed.
> As I wrote, I would like to have both: a pinned version for the PRs, so that the creator of the PR can concentrate on their own changes, and an unpinned version which shows the errors, but does not fail the build

That is a good point.  Hopefully quite rare, but I do remember that happening.  I'm on the fence as to whether it is worth the extra complication of having different versions in different workflows, vs. the occasional problem like that coming up.  Related point, I found it confusing doing the pydicom workflow changes recently, that the same thing is repeated over in several different jobs, in different workflow files.  I'd like to minimize that kind of duplication.

Reply:
I copied this to a [separate discussion issue](https://github.com/pydicom/pydicom/discussions/1815) unrelated to this PR - I propose to discuss this further over there.